# DeepSAP

**DeepSAP** is a Transformer-based workflow designed to enhance splice junction detection in RNA-seq data. By default, DeepSAP utilizes the highly sensitive GSNAP TGGA aligner for FASTQ inputs. Alternatively, it can also process pre-aligned BAM files generated by GSNAP directly. <br>

We evaluated the performance of DeepSAP in our article titled ***DeepSAP: A Novel Approach to Improve RNA-Seq Alignment Through Integration of Transcriptome Guidance and Transformer-Based Splice Junction Scoring***. In our benchmark, DeepSAP demonstrated an outstanding performance, achieving consistently the best results across all evaluated metrics using Baruzzo et al. datasets.

<img src="manuscript_data_code/Figure_3_benchmarking_data_results/All/Baruzzo_spirder_recallsVsprecision.png" width="1000">

For additional resources, including data, detailed analyses, and other supplementary materials related to the **DeepSAP** paper, please refer to the **paper_supplementals/README.md** file in this repository. 

## Table of Contents
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [Command-line Arguments](#command-line-arguments)
- [License/Terms of Use](#licenseterms-of-use)

## Requirements

- Docker with GPU support
- Input sorted alignment file in BAM/SAM or RNA-seq reads in FASTQ format
- Reference file in FASTA format
- Annotaion file in GTF format

## Installation
To use DeepSAP, you must have Docker with GPU support enabled and make sure the DeepSAP Docker image is available on your system. You can obtain the image by running the following command:
```bash
$ docker pull nvcr.io/nvidia/clara/clara-parabricks-deepsap:<TAG>
```

## Usage
You can use the accompanied dataset named **`malaria_short_pe`** under the [`test`](test/) folder to test DeepSAP's functionality with minimal setup. This dataset includes:

- Paired-end FASTQ files for alignment. 
- Malaria reference genome in FASTA format.
- Malaria annotation file in GTF format.

Update the paths in the provided example commands to point to the files in the [`test`](test/) folder.


1- Running DeepSAP with short-read RNA-seq FASTQ files

```bash
docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --rm   \
    --volume $(pwd)/test:/workdir                                                   \
    --volume $(pwd)/test/outputdir:/outputdir                                       \
    nvcr.io/nvidia/clara/clara-parabricks-deepsap:latest                            \
    --out /outputdir/                                                               \
    --prefix test_run_10K                                                           \
    --mate_1 /workdir/malaria_short_pe/SRR14793977_10K_1.fastq.gz                   \
    --mate_2 /workdir/malaria_short_pe/SRR14793977_10K_2.fastq.gz                   \
    --gtf /workdir/malaria_short_pe/Plasmodium_falciparum.ASM276v2.60.gtf           \
    --fasta /workdir/malaria_short_pe/Plasmodium_falciparum.ASM276v2.dna.toplevel.fa
```

2- Running DeepSAP with short-read RNA-seq FASTQ files and GSNAP index.

```bash
docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --rm   \
    --volume $(pwd)/test:/workdir                                                   \
    --volume $(pwd)/test/outputdir:/outputdir                                       \
    nvcr.io/nvidia/clara/clara-parabricks-deepsap:latest                            \
    --out /outputdir/                                                               \
    --prefix test_run_10K                                                           \
    --mate_1 /workdir/malaria_short_pe/SRR14793977_10K_1.fastq.gz                   \
    --mate_2 /workdir/malaria_short_pe/SRR14793977_10K_2.fastq.gz                   \
    --gtf /workdir/malaria_short_pe/Plasmodium_falciparum.ASM276v2.60.gtf           \
    --fasta /workdir/malaria_short_pe/Plasmodium_falciparum.ASM276v2.dna.toplevel.fa\
    --gsnap_idx /workdir/gsnap_idx/
```

<!-- 3- Running DeepSAP with an alignment BAM file generated from short-read RNA-seq data.

```bash
$docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 --rm \
    --volume $(pwd)/test:/workdir                               \
    --volume $(pwd)/test/outputdir:/outputdir                   \
    nvcr.io/nvidia/clara/clara-parabricks-deepsap:latest        \
    --out /outputdir/                                           \
    --prefix test_run                                           \
    --sam /workdir/short_read_pe_dataset/alignments.bam         \
    --fasta /workdir/short_read_pe_dataset/malaria_genome.fa    \
    --gtf /workdir/short_read_pe_dataset/malaria_annotation.gtf
``` -->

## Command-line Arguments

| Argument          | Description                                                                                          | Required       |
|-------------------|------------------------------------------------------------------------------------------------------|----------------|
| `-o, --out`       | Path to the output folder                                                                            | Yes            |
| `--prefix`        | Output files prefix string                                                                           | Yes            |
| `-g, --gtf`       | Path to the GTF annotation file compatible with the BAM file                                         | Yes            |
| `-f, --fasta`     | Path to the FASTA genome file compatible with the BAM file                                           | Yes            |
| `-s, --sam`       | Path to the SAM/BAM file or directory of files                                                       | Yes (if BAM)   |
| `--mate_1`        | Path to FASTQ file of mate 1 (for paired-end reads)                                                  | Yes (if FASTQ) |
| `--mate_2`        | Path to FASTQ file of mate 2 (for paired-end reads)                                                  | Yes (if FASTQ) |
| `--gsnap_idx`     | Path to GSNAP index                                                                                  | No             |
| `-c, --config`    | Config `.json` file to control DeepSAP internal parameters                                           | No             |
| `--batch`         | Batch size for inference                                                                             | No             |
| `--set_size`      | Set size to split datasets for inference                                                             | No             |
| `-t, --threads`   | Number of threads                                                                                    | No             |
| `--score_reads`   | Classify also reads using the transformer model and add scores to SAM, as appose to only SJ          | No             |
| `--n_reads`       | Number of reads to classify if `--score_reads` is used                                               | No             |
---


## License/Terms of Use
By pulling and using the Parabricks container, you accept the governing terms: The software and materials are governed by the NVIDIA Software License Agreement (found at https://www.nvidia.com/en-us/agreements/enterprise-software/nvidia-software-license-agreement/) and the Product-Specific Terms for NVIDIA AI Products (found at https://www.nvidia.com/en-us/agreements/enterprise-software/product-specific-terms-for-ai-products/); except for the model which is governed by the NVIDIA Models Community License Agreement(found at NVIDIA Community Model License). ADDITIONAL INFORMATION: Apache 2.0.